### ISSUE TITLE
scan fails the build with fail_build:false when testing with multiple devices

### ISSUE THREAD

**Issue by gobetti** [fastlane/fastlane#19827]
### New Issue Checklist

- [x] Updated fastlane to the latest version
- [x] I read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I read [docs.fastlane.tools](https://docs.fastlane.tools)
- [x] I searched for [existing GitHub issues](https://github.com/fastlane/fastlane/issues)

### Issue Description
My app builds fine during `scan` but if I have any failing unit tests, despite using `fail_build: false` the build still fails **if I specify `devices` as an array with length > 1**. Any commands I try to execute after `scan` in my custom lane will not be executed.

We have worked around this for years I believe but until fastlane 2.201.0 I didn't know it was specifically with multiple devices, but I've investigated more deeply now that our workaround stopped working.
The workaround was to call `scan` inside `begin` so we could `rescue` and manually call trainer (installed as a plugin) from `ensure`. Our trainer call now fails (#19826), and scan's internal call to trainer won't happen when any test fails since an exception is thrown before `handle_results` is reached in scan/runner.rb [as detailed in this discussion](https://github.com/fastlane/fastlane/pull/19629#discussion_r785388566)

##### Command executed
```
scan(
  devices: ['iPhone 13', 'iPhone 13 Pro Max'],
  fail_build: false
)
```

##### Complete output when running fastlane, including the stack trace and command used
Omitting the regular build output until the very end of the process where tests fail and so does `scan`:
<details>
+--------------------------------------+--------------------------------------+
|                          Summary for scan 2.201.0                           |
+--------------------------------------+--------------------------------------+
| devices                              | ["iPhone 13", "iPhone 13 Pro Max"]   |
| output_types                         | junit                                |
| fail_build                           | false                                |
| project                              | ./fastlane-201-issues.xcodeproj      |
| scheme                               | fastlane-201-issues                  |
| skip_detect_devices                  | false                                |
| ensure_devices_found                 | false                                |
| force_quit_simulator                 | false                                |
| reset_simulator                      | false                                |
| disable_slide_to_type                | true                                 |
| reinstall_app                        | false                                |
| clean                                | false                                |
| open_report                          | false                                |
| output_directory                     | ./fastlane/test_output               |
| buildlog_path                        | ~/Library/Logs/scan                  |
| include_simulator_logs               | false                                |
| xcodebuild_formatter                 | xcbeautify                           |
| output_remove_retry_attempts         | false                                |
| derived_data_path                    | /Users/gobetti/Library/Developer/Xc  |
|                                      | ode/DerivedData/fastlane-201-issues  |
|                                      | -ffyfswuojnhziqbryemguqagxyeg        |
| should_zip_build_products            | false                                |
| output_xctestrun                     | false                                |
| result_bundle                        | false                                |
| use_clang_report_name                | false                                |
| disable_concurrent_testing           | false                                |
| skip_build                           | false                                |
| slack_use_webhook_configured_userna  | false                                |
| me_and_icon                          |                                      |
| slack_username                       | fastlane                             |
| slack_icon_url                       | https://fastlane.tools/assets/img/f  |
|                                      | astlane_icon.png                     |
| skip_slack                           | false                                |
| slack_only_on_failure                | false                                |
| xcodebuild_command                   | env NSUnbufferedIO=YES xcodebuild    |
| skip_package_dependencies_resolutio  | false                                |
| n                                    |                                      |
| disable_package_automatic_updates    | false                                |
| use_system_scm                       | false                                |
| number_of_retries                    | 0                                    |
| xcode_path                           | /Applications/Xcode.app              |
+--------------------------------------+--------------------------------------+

...

Testing failed on 'iPhone 13'
2022-01-21 19:42:06.711 xcodebuild[53248:840539] [MT] IDETestOperationsObserverDebug: 24.863 elapsed -- Testing started completed.
2022-01-21 19:42:06.711 xcodebuild[53248:840539] [MT] IDETestOperationsObserverDebug: 0.000 sec, +0.000 sec -- start
2022-01-21 19:42:06.711 xcodebuild[53248:840539] [MT] IDETestOperationsObserverDebug: 24.863 sec, +24.863 sec -- end
Testing failed on 'iPhone 13 Pro Max'
Failing tests:
	fastlane-201-issuesTests:
		fastlane_201_issuesTests.testExample()

** TEST FAILED **
[19:42:12]: Exit status: 65
[19:42:12]: ❌ scan failed, exception backtrace: ["/Users/gobetti/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.201.0/fastlane_core/lib/fastlane_core/ui/interface.rb:163:in `build_failure!'", "/Users/gobetti/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.201.0/fastlane_core/lib/fastlane_core/ui/ui.rb:17:in `method_missing'", "/Users/gobetti/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.201.0/scan/lib/scan/error_handler.rb:31:in `handle_build_error'", "/Users/gobetti/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.201.0/scan/lib/scan/runner.rb:92:in `block in execute'", "/Users/gobetti/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.201.0/fastlane_core/lib/fastlane_core/command_executor.rb:84:in `execute'", "/Users/gobetti/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.201.0/scan/lib/scan/runner.rb:79:in `execute'", "/Users/gobetti/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.201.0/scan/lib/scan/runner.rb:55:in `test_app'", "/Users/gobetti/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.201.0/scan/lib/scan/runner.rb:23:in `run'", "/Users/gobetti/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.201.0/scan/lib/scan/manager.rb:23:in `work'", "/Users/gobetti/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.201.0/fastlane/lib/fastlane/actions/run_tests.rb:16:in `run'", "/Users/gobetti/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.201.0/fastlane/lib/fastlane/runner.rb:263:in `block (2 levels) in execute_action'", "/Users/gobetti/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.201.0/fastlane/lib/fastlane/actions/actions_helper.rb:69:in `execute_action'", "/Users/gobetti/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.201.0/fastlane/lib/fastlane/runner.rb:255:in `block in execute_action'", "/Users/gobetti/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.201.0/fastlane/lib/fastlane/runner.rb:229:in `chdir'", "/Users/gobetti/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.201.0/fastlane/lib/fastlane/runner.rb:229:in `execute_action'", "/Users/gobetti/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.201.0/fastlane/lib/fastlane/runner.rb:157:in `trigger_action_by_name'", "/Users/gobetti/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.201.0/fastlane/lib/fastlane/fast_file.rb:159:in `method_missing'", "Fastfile:23:in `block (2 levels) in parsing_binding'", "/Users/gobetti/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.201.0/fastlane/lib/fastlane/lane.rb:33:in `call'", "/Users/gobetti/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.201.0/fastlane/lib/fastlane/runner.rb:49:in `block in execute'", "/Users/gobetti/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.201.0/fastlane/lib/fastlane/runner.rb:45:in `chdir'", "/Users/gobetti/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.201.0/fastlane/lib/fastlane/runner.rb:45:in `execute'", "/Users/gobetti/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.201.0/fastlane/lib/fastlane/lane_manager.rb:47:in `cruise_lane'", "/Users/gobetti/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.201.0/fastlane/lib/fastlane/command_line_handler.rb:36:in `handle'", "/Users/gobetti/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.201.0/fastlane/lib/fastlane/commands_generator.rb:109:in `block (2 levels) in run'", "/Users/gobetti/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/commander-4.6.0/lib/commander/command.rb:187:in `call'", "/Users/gobetti/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/commander-4.6.0/lib/commander/command.rb:157:in `run'", "/Users/gobetti/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/commander-4.6.0/lib/commander/runner.rb:444:in `run_active_command'", "/Users/gobetti/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.201.0/fastlane_core/lib/fastlane_core/ui/fastlane_runner.rb:124:in `run!'", "/Users/gobetti/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/commander-4.6.0/lib/commander/delegates.rb:18:in `run!'", "/Users/gobetti/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.201.0/fastlane/lib/fastlane/commands_generator.rb:353:in `run'", "/Users/gobetti/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.201.0/fastlane/lib/fastlane/commands_generator.rb:42:in `start'", "/Users/gobetti/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.201.0/fastlane/lib/fastlane/cli_tools_distributor.rb:122:in `take_off'", "/Users/gobetti/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/fastlane-2.201.0/bin/fastlane:23:in `<top (required)>'", "/Users/gobetti/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/bin/fastlane:23:in `load'", "/Users/gobetti/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/bin/fastlane:23:in `<top (required)>'", "/Users/gobetti/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/bundler-2.1.4/lib/bundler/cli/exec.rb:63:in `load'", "/Users/gobetti/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/bundler-2.1.4/lib/bundler/cli/exec.rb:63:in `kernel_load'", "/Users/gobetti/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/bundler-2.1.4/lib/bundler/cli/exec.rb:28:in `run'", "/Users/gobetti/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/bundler-2.1.4/lib/bundler/cli.rb:476:in `exec'", "/Users/gobetti/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/bundler-2.1.4/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'", "/Users/gobetti/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/bundler-2.1.4/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'", "/Users/gobetti/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/bundler-2.1.4/lib/bundler/vendor/thor/lib/thor.rb:399:in `dispatch'", "/Users/gobetti/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/bundler-2.1.4/lib/bundler/cli.rb:30:in `dispatch'", "/Users/gobetti/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/bundler-2.1.4/lib/bundler/vendor/thor/lib/thor/base.rb:476:in `start'", "/Users/gobetti/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/bundler-2.1.4/lib/bundler/cli.rb:24:in `start'", "/Users/gobetti/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/bundler-2.1.4/exe/bundle:46:in `block in <top (required)>'", "/Users/gobetti/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/bundler-2.1.4/lib/bundler/friendly_errors.rb:123:in `with_friendly_errors'", "/Users/gobetti/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/bundler-2.1.4/exe/bundle:34:in `<top (required)>'", "/Users/gobetti/.rbenv/versions/2.6.5/bin/bundle:23:in `load'", "/Users/gobetti/.rbenv/versions/2.6.5/bin/bundle:23:in `<main>'"]

+------+------------------+-------------+
|           fastlane summary            |
+------+------------------+-------------+
| Step | Action           | Time (in s) |
+------+------------------+-------------+
| 1    | default_platform | 0           |
| 💥   | scan             | 37          |
+------+------------------+-------------+
</details>

### Environment
It doesn't seem relevant here, any environment should be able to reproduce this, but to make this easier I created a repository https://github.com/gobetti/fastlane-201-issues where you can use the branch `fail-build-multiple-devices`. Or this commit https://github.com/gobetti/fastlane-201-issues/commit/3a8a631fe35a71c0612a66de8f811f83882ed000
This other commit further proves that an exception is thrown from scan by printing from inside `rescue` https://github.com/gobetti/fastlane-201-issues/commit/ac560ca85c94b72133f6f7907925f0598638b0ae
That repository contains a vanilla Xcode 13 project, my only change to it was to add a failing unit test.

**Comment by joshdholtz** [comment]
Will look into this!

**Comment by joshdholtz** [comment]
@gobetti Got a fix for this one! This one was really fun to track down 😛 But thank you soooo much for that awesome example project repo. That was actually a **HUGE** help 🥰  Fix is in #19829

**Comment by gobetti** [comment]
I suspect example projects do allow you to much better focus your time on fixing, also the git history helps proving where an issue is introduced. Could possibly suggest that over requesting huge console output and environment logs - it must be no fun to parse those 😬
so this was the least I could do with my limited knowledge when it comes to fixing fastlane in a way that doesn't break the tool for millions of users. Thank you so much for looking into this and fixing so quickly! 🎉
